server/src/main/java/org/opensearch/common/blobstore/stream/read/listener/ReadContextListener.java

```diff
@@ -10,56 +10,155 @@
-    private final ThreadPool threadPool;
     private final ActionListener<String> completionListener;
-    private static final Logger logger = LogManager.getLogger(ReadContextListener.class);
+    private final ThreadPool threadPool;
+    private final UnaryOperator<InputStream> rateLimiter;
 
-    public ReadContextListener(String fileName, Path fileLocation, ThreadPool threadPool, ActionListener<String> completionListener) {
-        this.fileName = fileName;
+    public ReadContextListener(
+        String blobName,
+        Path fileLocation,
+        ActionListener<String> completionListener,
+        ThreadPool threadPool,
+        UnaryOperator<InputStream> rateLimter
+    ) {
+        this.blobName = blobName;
         this.fileLocation = fileLocation;
-        this.threadPool = threadPool;
         this.completionListener = completionListener;
+        this.threadPool = threadPool;
+        this.rateLimiter = rateLimter;
     }
 
     @Override
     public void onResponse(ReadContext readContext) {
-        logger.trace("Streams received for blob {}", fileName);
+        logger.debug("Received {} parts for blob {}", readContext.getNumberOfParts(), blobName);
         final int numParts = readContext.getNumberOfParts();
-        final AtomicBoolean anyPartStreamFailed = new AtomicBoolean();
-        FileCompletionListener fileCompletionListener = new FileCompletionListener(numParts, fileName, completionListener);
-
-        for (int partNumber = 0; partNumber < numParts; partNumber++) {
-            FilePartWriter filePartWriter = new FilePartWriter(
-                partNumber,
-                readContext.getPartStreams().get(partNumber),
-                fileLocation,
-                anyPartStreamFailed,
-                fileCompletionListener
-            );
-            threadPool.executor(ThreadPool.Names.GENERIC).submit(filePartWriter);
+        final AtomicBoolean anyPartStreamFailed = new AtomicBoolean(false);
+        final GroupedActionListener<String> groupedListener = new GroupedActionListener<>(
+            ActionListener.wrap(r -> completionListener.onResponse(blobName), completionListener::onFailure),
+            numParts
+        );
+        final Queue<ReadContext.StreamPartCreator> queue = new ConcurrentLinkedQueue<>(readContext.getPartStreams());
+        final StreamPartProcessor processor = new StreamPartProcessor(
+            queue,
+            anyPartStreamFailed,
+            fileLocation,
+            groupedListener,
+            threadPool.executor(ThreadPool.Names.REMOTE_RECOVERY),
+            rateLimiter
+        );
+        // We have a dedicated thread pool for draining streams and writing them to a file on the
+        // local disk. However, we want more concurrent requests to fetch stream parts to be
+        // happening asynchronously so that there is always a queue of streams ready to be consumed
+        // by the REMOTE_RECOVERY thread pool.
+        final int concurrentStreamRequests = 20;
```

Comment:
This constant isn't ideal. In my limiting testing this value seems to work pretty well. I'm not exactly sure how to parameterize or scale this value, so just starting simple with a constant value. The underlying connection pool for the S3 repository has a default size of 100 (also not scaled by host size).

Reply:
can we make this dynamic with default size of 20 ?

Reply:
+1 on making this tunable

Reply:
Added a new dynamic setting (`indices.recovery.max_concurrent_remote_store_streams`). Unfortunately it was not super simple to wire in and I'm not totally happy with how I did it, but this seems like the best way absent a much larger refactoring.